cmd/snap/cmd_run.go

```diff
@@ -493,30 +483,90 @@ func (x *cmdRun) straceOpts() (opts []string, raw bool, err error) {
 	return opts, raw, nil
 }
 
+func isPossibleRaceCondition(oldInfo *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock) bool {
+	if !features.RefreshAppAwareness.IsEnabled() || app.IsService() || hintFlock != nil {
+		// Skip check
+		return false
+	}
+
+	currentInfo, err := snap.ReadCurrentInfo(oldInfo.InstanceName())
+	if err != nil {
+		// Current symlink does not exist, snap is either removed or being refreshed.
+		return true
+	}
+	if currentInfo.SnapRevision() != oldInfo.SnapRevision() {
+		// Snap was refreshed while we did not hold the hint lock.
+		return true
+	}
+
+	return false
+}
+
```

Comment:
@olivercalder @pedronis 
I believe this check could be simplified to `oldHintLock == nil && Exists(hintFile)`, I think it would be even more robust.

We started without a hint file and now we do, let's retry just in case.